crates/alloy-dyn-contract/src/interface.rs

```diff
@@ -0,0 +1,123 @@
+use std::collections::HashMap;
+
+use crate::{
+    error::{Error, Result},
+    instance::ContractInstance,
+};
+use alloy_dyn_abi::{DynSolValue, FunctionExt, JsonAbiExt};
+use alloy_json_abi::{Function, JsonAbi};
+use alloy_primitives::{Address, Selector};
+
+/// A smart contract interface.
+#[derive(Debug, Clone)]
+pub struct Interface {
+    abi: JsonAbi,
+    functions: HashMap<Selector, (String, usize)>,
+}
+
+impl From<JsonAbi> for Interface {
+    fn from(abi: JsonAbi) -> Self {
+        Self { abi, functions: Default::default() }
+    }
+}
+
+// TODO: events/errors
```

Comment:
just flagging :D

Reply:
yea we don't have the facilities rn for events so it's more of a follow up to do

Reply:
sweet np, let's open an issue for this and roll